libraries/stdlib/osl/stdosl.h

```diff
@@ -573,6 +573,169 @@ closure color volume_matte() BUILTIN;
+// inside the medium.
+//
+// TODO:
+// - Transmission handling for this node has not been fully defined yet in MaterialX.
+//   In particular how is IOR for refractions derived from the f0, f90 parameterization?
+//   Do we just derive it from f0? For an artist it seems hard to control refractions that way.
+//
+closure color generalized_schlick_bsdf(normal N, vector U, color reflection_tint, color transmission_tint, float roughness_x, float roughness_y, color f0, color f90, float exponent, string distribution) BUILTIN;
+​
+// Constructs a translucent (diffuse transmission) BSDF based on the Lambert reflectance model.
+//
+closure color translucent_bsdf(normal N, color albedo) BUILTIN;
+
+// Constructs a closure that represents straight transmission through a surface.
+//
+// NOTE:
+//  - This is not a node in the MaterialX library, but the surface shader constructor
+//    node has an 'opacity' parameter to control textured cutout opacity.
+//
+closure color transparent_bsdf() BUILTIN;
+​
+// Constructs a BSSRDF for subsurface scattering within a homogeneous medium.
+//
+// TODO:
+// - Do we want/need this dedicated BSSRDF closure? We have had discussions before about
+//   describing this more rigorously as a BSDF layered over a VDF instead.
+//   In the MaterialX library we want this as a node for the user friendly interface,
+//   but for back-end implementation maybe a dedicated closure is not needed?
+//
+closure color subsurface_bssrdf(normal N, color albedo, color radius, float anisotropy) BUILTIN;
+​
+// Constructs a microfacet BSDF for the back-scattering properties of cloth-like materials.
+// This closure may be vertically layered over a base BSDF, where energy that is not reflected
+// will be transmitted to the base closure.
+//
+closure color sheen_bsdf(normal N, color albedo, float roughness) BUILTIN;
+​
+// Adds an iridescent thin film layer over a microfacet base BSDF. This must be layered over
+// another base BSDF using the layer() closure, as this is a modifier and cannot be used as a
+// standalone closure.
+//
+// TODO:
+// - This might be better to represent as optional extra arguments on the closures that do support
+//   thin-film iridescence (dielectric_bsdf, conductor_bsdf and generalized_schlick_bsdf)?
+//
+closure color thin_film_bsdf(float thickness, float ior) BUILTIN;
+​
+​
+// -------------------------------------------------------------//
+// EDF closures                                                 //
+// -------------------------------------------------------------//
+​
+// Constructs an EDF emitting light uniformly in all directions.
+//
+closure color uniform_edf(color emittance) BUILTIN;
+
+// Constructs an EDF emitting light inside a cone around the normal direction.
+//
+closure color conical_edf(color emittance, normal N, float inner_angle, float outer_angle) BUILTIN;
+
+// Constructs an EDF emitting light according to a measured IES light profile.
```

Comment:
IES profiles are usually defined for point light sources. By attaching the IES profile within the shader, it implicitly becomes a property of a specific point on the surface which has the effect of "blurring" the profile. Is this what is intended?

It also makes the light potentially quite a bit harder to sample. Some techniques for efficient IES profile sampling assume a constant profile across the surface.

I'm curious if any implementation are supporting this node already.

Reply:
Yes, this EDF was included in the PBS library mainly for constructing light shaders. In MaterialX there is a concept of light shaders, where for a point light it could be described as a light shader attached to a locator, and not necessarily a surface. I suppose you could use this EDF to a surface shader as well, but yes then it would be blurring the profile over the surface. The only implementation supporting this node as of now is Nvidia MDL which has a dedicated EDF closure for IES profiles.

I'm not sure OSL was intended to be used for light shaders of local lights in this manner? If not we can remove this for now. In any case the use of this EDF would be very rare and it can definitely be placed at the bottom of the backlog.

Reply:
In the beginning we thought of using OSL for light shading too, and we tried. But we quickly saw it wasn't practical for us and we just define plain emission in the shader. The render is responsible for profiles.